src/vs/workbench/api/common/extHost.protocol.ts

```diff
@@ -1086,8 +1086,7 @@ export interface MainThreadWorkspaceShape extends IDisposable {
 	$updateWorkspaceFolders(extensionName: string, index: number, deleteCount: number, workspaceFoldersToAdd: { uri: UriComponents; name?: string }[]): Promise<void>;
 	$resolveProxy(url: string): Promise<string | undefined>;
 	$requestWorkspaceTrust(options?: WorkspaceTrustRequestOptions): Promise<boolean | undefined>;
-	$registerEditSessionIdentityProvider(scheme: string): void;
-	$unregisterEditSessionIdentityProvider(scheme: string): void;
+	$registerEditSessionIdentityProvider(scheme: string): Promise<IDisposable>;
```

Comment:
Oh, my feedback was a little too short/confusing. On the "wire" you need to use the register/unregister pair because you cannot return an `IDisposable` over it. However, the service-layer should use a disposable. What I usually do the following

* service has register-call which return a disposable to unregister
* API $register is called with a "random/monotonic" handle
* API stores the disposable by this handle
* API $unregister is called with the handle again - to lookup and call the respective disposable

This is a good sample https://github.com/microsoft/vscode/blob/0656d21d11910e1b241d7d6c52761d89c0ba23a4/src/vs/workbench/api/browser/mainThreadDecorations.ts#L84-L106